tests/plugins/Qukeys/basic/test/testcase.cpp

```diff
@@ -0,0 +1,164 @@
+/* -*- mode: c++ -*-
+ * Copyright (C) 2020  Keyboard.io, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "testing/setup-googletest.h"
+
+#include "../sketch.h"
+
+SETUP_GOOGLETEST();
+
+namespace kaleidoscope {
+namespace testing {
+namespace {
+
+struct QukeyValues {
+  Key primary{Key_NoKey};
+  Key alternate{Key_NoKey};
+};
+constexpr KeyAddr key_addr_A{2, 1};
+constexpr KeyAddr key_addr_S{2, 2};
+constexpr KeyAddr key_addr_D{2, 3};
+constexpr KeyAddr key_addr_F{2, 4};
+constexpr KeyAddr key_addr_X{3, 2};
+
+using ::testing::IsEmpty;
+
+class QukeysBasic : public VirtualDeviceTest {};
+
+TEST_F(QukeysBasic, TapQukeyAlone) {
+
+  std::unique_ptr<State> state{nullptr};
+  std::set<uint8_t> expected_keycodes{};
+
+  // Press `A`
+  sim_.Press(key_addr_A);
+
+  state = VirtualDeviceTest::RunCycle();
+
+  ASSERT_EQ(state->HIDReports()->Keyboard().size(), 0);
+
+  // Release `A`
+  sim_.Release(key_addr_A);
+  expected_keycodes.insert(Key_A.getKeyCode());
+
+  state = VirtualDeviceTest::RunCycle();
+
+  ASSERT_EQ(state->HIDReports()->Keyboard().size(), 1);
+  EXPECT_THAT(state->HIDReports()->Keyboard(0).ActiveKeycodes(),
+              ::testing::ElementsAreArray(expected_keycodes));
+
+  expected_keycodes.erase(Key_A.getKeyCode());
+
+  for (int i{0}; i < 10; ++i) {
```

Comment:
int i{0} is a very uncommon way to initialize ints. I see that you do it in QuKeys, but it's not a convention we use anywhere else in the codebase that I'm aware of. 

When I saw it on your constexprs, I had to stop and try to figure out what's going on. 

 I'd love to hear more about why you've used this style.

Reply:
`int i{0}` is C++ list initialization. It's something I picked up via conversations with Florian, and I've been using it ever since. It avoids at least one problem where the compiler parses something as a function declaration instead of a variable initialization via a constructor. There are a number of articles on "[the most vexing parse](https://www.fluentcpp.com/2018/01/30/most-vexing-parse/)". With braces, that problem is avoided, and it's possible to consistently use that style to initialize all variables in C++, even primitives.

That said, I'm not going to argue about it if you want to use `=` in cases like this. I just find it useful as a marker that a) I'm looking at/writing C++, not C, and b) this is a variable initialization, not merely assignment. Of course, this breaks down when using `auto`…